### ISSUE TITLE
python's SimpleHTTPServer does not work anymore as backend

### ISSUE THREAD

**Issue by Geal** [sozu-proxy/sozu#109]
log:
```
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 54413)
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 290, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 318, in process_request
    self.finish_request(request, client_address)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 331, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 654, in __init__
    self.finish()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/SocketServer.py", line 713, in finish
    self.wfile.close()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/socket.py", line 283, in close
    self.flush()
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/socket.py", line 307, in flush
    self._sock.sendall(view[write_offset:write_offset+buffer_size])
error: [Errno 32] Broken pipe
```

**Comment by BlackYoup** [comment]
I've poke around the issue, still not sure why this fails from time to time but what I found is that it fails in the `match` here: https://github.com/sozu-proxy/sozu/blob/master/lib/src/network/protocol/http.rs#L456

When it fails, the `self.state.response` is `Some(HasStatusLine(RStatusLine { version: "10",status: 200, reason: "OK" }, Close))`
When it works, it's: `Some(ResponseWithBody(RStatusLine { version: "10", status: 200, reason: "OK" }, Close, 112))`

And the `ClientResult::CloseBothFailure` is returned at the end of the match

**Comment by Geal** [comment]
reactivate the log here: https://github.com/sozu-proxy/sozu/blob/master/lib/src/parser/http11.rs#L1338
and try again. It will be slow, but we'll see what fails to parse

**Comment by BlackYoup** [comment]
After re-activating those logs, I didn't notice anything, once the buffer is fully consumed, it just ends the loop. I print the content of `buf.buffer.memory` and it seems like it doesn't contain much more of what's being parsed right after.
So it seems like the problem is not in the parsing itself, but more likely in the network layer since the buffer doesn't hold the full server's response (and maybe the server didn't sent it)

**Comment by BlackYoup** [comment]
The python server fully respond to the proxy.

Here is a tcpdump output: https://gist.github.com/BlackYoup/476651c207c8ef19f088509200ff7c59
Curl output: https://gist.github.com/BlackYoup/01326c64a5436ef568d49308c54ab872

I'm starting to dive into totally unknown waters from here

**Comment by BlackYoup** [comment]
All of this took me here: https://github.com/sozu-proxy/sozu/blob/master/lib/src/network/socket.rs#L28

When it fails, the `.read()` action returns an `ErrorKind::WouldBlock` error. When the proxy successfully transmit the whole data, this is the `Ok(0)` line that's being encountered.

My guess here is that the loop is too fast and the python server don't have the time to send the remaining data until the next socket `read`. I added a `sleep(Duration::from_millis(10));` at the start of the loop and I can't reproduce the problem anymore :-)

Does it seem logic / possible ?

**Comment by Geal** [comment]
Fixed in 134a9f1. The state machine was assuming that when the parser says "we can restart parsing" from header parsing, it means all the headers were parsed. Now it waits for the rest (even if it started sending the headers to the other end).

**Comment by BlackYoup** [comment]
The problem happens again, at least I can reproduce on two machines. Adding the delay in the same loop as my comment above fixes it.

This time, the client only receives the response status line (HTTP/1 200 OK). Python first write the status line and then each header in distinct packets. At the end, the body is written. On my tests:

Tcpdump summary
```
Time                  Content

00:12:53.503670       Status line
00:12:53.503727       Server header
00:12:53.503771       Date header
00:12:53.503805       Content-Type header
00:12:53.503844       Content-Length header
00:12:53.503918       Body
```

I did a bisect between 134a9f1 (your fix above) and 93a12d34bce5a05b4c5f200f9f02ba2f5bde1aca (current master) and the commit returned is 7ebdddd3abe271e6bbad368e6c52c6e5b0cdb34a

Let me know if you need more details
